### ISSUE TITLE
The AD path still truncates the CTM chi bond by the old per-sector quota

### ISSUE THREAD

**Issue by yingjerkao** [tenax-lab/tenax#929]
> 🤖 **AI-generated issue** — written by Claude Code, posted by @Ying-Jer Kao.

Split out of #922, which fixed the CTM chi-bond truncation on the **eager**
path. The AD path still carries the defect, because the fix is not expressible
there.

## What #922 changed, and where it stops

`base_charges` used to be tiled by `_derive_charges` into a per-sector *quota*:
every sector capped at its tiled share, and any charge absent from
`base_charges` allocated a share of zero. #922 replaced that with the global
top-chi plus a one-slot floor per named charge, and the symmetric CTM now
tracks the dense reference to 2.2e-16 at D=2 chi=24.

That cut reads singular values. Under `jax.jit` there are none to read: the
per-sector block shapes are baked at trace time, so which sector owns each chi
slot has to be decided *before* the SVD runs.
`linalg._truncated_svd_symmetric_traced` therefore still applies the quota, and
`optimize_gs_ad` still truncates the old way.

The divergence is pinned, not silent:
`test_the_traced_path_pins_the_chi_new_inventory` asserts eager keeps
`{0: 3, 1: 1}` where traced keeps `{0: 2, 1: 2}` on the projector fixture.

## Two static replacements were measured, and both are worse somewhere

| static rule | D=2 chi=24 | D=3 chi=8 |
|---|---|---|
| tiled quota (shipped today) | 2.58e-03 | 1.11e-16 |
| capacity-proportional + floor | 6.26e-14 (chi=16) | 1.08e-05 |
| eager value-aware cut (#922) | 2.22e-16 | 2.22e-16 |

Neither proxy tracks the spectrum. Capacity fixes D=2 by giving `|q|=4` its
share, and breaks D=3 chi=8 by spending a slot on `q=-4` that the spectrum says
is worthless — 11 orders. This is why #922 shipped neither and left
`linalg.svd` untouched (which also leaves fPEPS simple update's pinned layout
alone, #558).

## Suggested design

The inventory has to come from outside the trace. The obvious candidate is the
**incoming chi leg's own charges**: `TensorIndex.charges` is static metadata
even under `jit`, the leg is already in scope in
`_compute_2x2_projector_symmetric`, and at a fixed point its multiset *is* what
the eager cut converged to — which is precisely the regime CTM-AD linearises
around. The first sweep would reproduce the initial env's inventory, as today.

Open questions before building it:

1. Which of the two chi legs of `Q_TL` names the seam being compressed, per
   direction (`left`/`right`/`top`/`bottom`)?
2. What happens when a sector cannot supply its share — the existing greedy
   fill, or something that preserves the multiset exactly?
3. Does the AD energy actually move? Measure `optimize_gs_ad` on a D=2 and a
   D=3 U(1)-Sz pair against the eager-CTM energy at the same chi before
   building anything: if the gap is already at the optimiser's noise floor,
   this is not worth the plumbing.

Gate on (3) first.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @Ying-Jer Kao.

Second item for this issue, found by Codex review on #928 and confirmed pre-existing.

**`_svd_projector_symmetric` drops a named base charge whose grown corners give `M_q.size == 0`.** That sector never reaches `sector_results`, so it contributes no singular value and the #922 floor has nothing to reserve — the charge leaves the bond index entirely, which `contract()` cannot undo. `main` behaves the same way (the `_derive_charges` quota it replaced had no `else` branch), so this is not a #928 regression, and the default 2x2 recipe does not route through this function.

It was left unfixed because the obvious fix is not obviously right:

- `_eigh_projector_symmetric` seeds an identity column, which works because the eigh projector *is* the eigenvector matrix.
- The SVD projector is `P1 = C4g @ V S^{-1/2}`, built from corner data. With `S = 0` a seeded slot yields a **zero** projector column; with `S = 1` it injects a fake largest singular value that reorders the truncation and misreports `eps_T`.

So whoever picks this up needs to decide what a projector column for a sector with no data should *be*, and measure it — a zero column is defensible under #905's "allocated-but-zero is recoverable, absent is not" reading, but it makes the projector rank-deficient and that has not been tested.

The documented limit and its guard are in `test_the_floor_cannot_conjure_a_sector_the_decomposition_omits`.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @Ying-Jer Kao.

**Correcting this issue: its central claim is wrong, and #930's remaining scope is much smaller than I wrote.**

I went to build the second half — seeding `env_init` from an eager pre-pass so a cold AD run inherits a spectrum-derived inventory — and the measurement refused the premise before any code was written.

## `ctm_ad_mode="root_implicit_symmetric"` was never affected

This issue says "`optimize_gs_ad` still carries #922's truncation". That is wrong for the documented symmetric optimiser. Spying on `_compute_2x2_projector_symmetric` across a full `sym_root_implicit_energy_and_grad` call counts **zero invocations**, eager or traced — that path has its own sweep. And its truncation, in `_ctm_root_implicit_sym_sectors.sector_svd`, is:

```python
    # Global truncation across sectors.
    ranked = sorted(((float(sv), q, i) for q, (_U, s, *_r) in raw.items()
                     for i, sv in enumerate(s)), key=lambda t: -t[0])
    dims = dict.fromkeys(raw, 0)
    for _sv, q, _i in ranked[: int(chi)]:
        dims[q] += 1
    layout = BondLayout.from_dims(dims)
```

One global top-χ over the union of the sector spectra — exactly #922's policy — recorded in a `BondLayout` that is deliberately *frozen at the converged point* so the adjoint solves the same-sized system the forward built. That is the architecture I proposed in the "suggested design" above, already implemented, and better than what I proposed.

So the seeding work is **not warranted**. It would plumb an inventory into an `env_cache` that the symmetric optimiser does not use.

## What #930 does fix is real, and reachable

The traced 2×2 projector is reached by the *other* symmetric AD entry points: symmetric site tensors through `ctm_energy_explicit` under `jax.grad` take it **8 times** in a 2-sweep warmup (eagerly zero), and the gradient comes back finite. #930 makes those inherit the environment's inventory instead of re-imposing a tiled guess every sweep.

What remains for those entry points is only the *cold* inventory: `initialize_ctm_tensor_env` still tiles, and inheritance perpetuates it. That is a much narrower residue than this issue claims, and it is not on the path anyone optimising a symmetric state actually takes.

## Recorded for whoever picks the residue up

The inventory settles after **2 eager sweeps** — at both D=2 χ=8 and D=3 χ=16 the 2-sweep inventory equals the 200-sweep one (7.1s and 7.6s, against 503s and 229s to converge). So a seeding pre-pass, if it is ever wanted, is cheap.

#930 adds a guard pinning `sector_svd`'s global truncation on a deliberately lopsided fixture, so the root-implicit path cannot quietly regress into a per-sector quota the way the eager cut had.

**Comment by yingjerkao** [comment]
> 🤖 **AI-generated comment** — written by Claude Code, posted by @Ying-Jer Kao.

Closing. What was actionable here landed in #930 (`589f77e`); what is left is being declined deliberately, and this records why so it is not rediscovered as an open lead.

**The premise was wrong.** This issue said "`optimize_gs_ad` still carries #922's truncation". That is false for the documented symmetric optimiser: `ctm_ad_mode="root_implicit_symmetric"` does not use the 2×2 projector at all — measured, **zero** invocations across a full `sym_root_implicit_energy_and_grad` call — and `_ctm_root_implicit_sym_sectors.sector_svd` already takes one global top-χ over the union of the sector spectra, recording it in a `BondLayout` frozen for the adjoint. That is the design this issue proposed under "suggested design", already implemented and better than what was proposed. #930 added a guard so it cannot regress into a quota the way the eager cut did.

**What #930 fixed.** The traced 2×2 projector is reachable — symmetric site tensors through `ctm_energy_explicit` under `jax.grad` take it 8 times in a 2-sweep warmup, eagerly zero, gradient finite. It now inherits the environment's own χ inventory instead of re-imposing the double-layer charges tiled to χ. Seeding the static rule with the eager inventory reproduces the eager environment *exactly* (`|E_eager − E_seeded| = 0.0` at D=2 χ=8 and D=3 χ=16) and does so faster, since a fixed inventory does not churn block shapes.

**What is declined.** A *cold* run of those two entry points still starts from `initialize_ctm_tensor_env`'s tiled inventory, and inheritance perpetuates it. Closing that means seeding `env_init` from an eager pre-pass — three call sites in `ipeps_optimize.py`, χ-ramp interaction, and full optimiser runs to verify — for a path that is not the one anyone optimising a symmetric state takes. The cost is not justified by the benefit.

**Recorded for whoever disagrees.** The pre-pass would be cheap: the χ inventory is already final after **2 eager sweeps** — the 2-sweep inventory equals the 200-sweep one at both D=2 χ=8 and D=3 χ=16 (7.1 s and 7.6 s, against 503 s and 229 s to converge). So if this is ever wanted, the expensive-looking part isn't.
